grpc/src/main/java/io/stargate/grpc/service/SchemaNotificationsHandler.java

```diff
@@ -35,6 +35,7 @@
 
   private final Persistence persistence;
   private final ServerCallStreamObserver<SchemaNotification> responseObserver;
+  private boolean isFailed;
```

Comment:
Just unregistering in `setOnCloseHandler` did not completely get rid of the duplicate calls, maybe there is a race condition that still allows another event to be processed.

So I had to add this to explicitly check from our code. It doesn't add much complexity since we can piggyback on the existing synchronization.

Reply:
This looks heavily to the issues that we were seeing with the bi-directional streaming.. Is this a streaming handler? If so you can not call `onError` once you hit the exception.. You rather need to send the exception back as the result of the call.. 

You must call `onError` only once.. This however, kills the client connection and thus you can not continue streaming the responses..

Maybe I am completely wrong as I am jumping here without a context, but if you have duplicate calls for the `onError` it's probably something done wrong.. And I don't think the solution in the PR is actually solving the root issue..

Reply:
+1 to @ivansenic comment. We cannot call `onError` twice, according to the contract it can be called only once.
Please take a look how I solved it in:
https://github.com/stargate/stargate/pull/1469 and the wrapper type:
https://github.com/stargate/stargate/pull/1469/files#diff-edefd35d46c51b4410ea5494061770729922d6fa5548f158078c5e4fd37a35f4R356

Reply:
> We cannot call `onError` twice

Agreed, this is what the PR fixes.

> You rather need to send the exception back as the result of the call..

I've considered it but I don't think it's the right approach here. Errors are unexpected here. It's not like bidirectional streaming where an exception can be a legitimate response to one of the queries. If the code throws it's a sign that something is seriously wrong, so I think failing the observer is the right decision, it's up to the client to decide if it wants to resubscribe to notifications.

> And I don't think the solution in the PR is actually solving the root issue..

Fair enough, the fix ensures that extra calls to `onError` are ignored, but I didn't get to the bottom of why they happen in the first place.
What should happen is that after the first error, the callback from `setOnCloseHandler` is executed, we unregister from the persistence and therefore `onSchemaChange` should not happen again. As alluded above I suspect there is a race condition that allows multiple events to go through before we are unregistered.

I'll push a commit with debug logs to try to get more details.

Reply:
The error happens when a keyspace that contained a table gets dropped.
1. We get a notification for the table:
    ```
    INFO  [MigrationStage:1] 2022-01-20 17:22:44,659 onSchemaChange(DROPPED, TABLE, ks1, tbl1)
    ```
2. we try to build the keyspace description to include in the notification. But the keyspace is already gone so this throws, and we notify the observer of the error:
    ```
    INFO  [MigrationStage:1] 2022-01-20 17:22:44,662 invoke responseObserver.onError
    io.grpc.StatusException: NOT_FOUND: Keyspace not found
      at io.grpc.Status.asException(Status.java:543)
      at io.stargate.grpc.service.SchemaHandler.buildKeyspaceDescription(SchemaHandler.java:66)
      at io.stargate.grpc.service.SchemaNotificationsHandler.onSchemaChange(SchemaNotificationsHandler.java:79)
    ```
3. then we get a notification for the keyspace:
    ```
    INFO  [MigrationStage:1] 2022-01-20 17:22:44,663 onSchemaChange(DROPPED, KEYSPACE, ks1, null)
    ```
4. we try to notify the observer, but this fails because it's already complete:
    ```
    INFO  [MigrationStage:1] 2022-01-20 17:22:44,663 invoke responseObserver.onNext
    ...
    INFO  [MigrationStage:1] 2022-01-20 17:22:44,663 invoke responseObserver.onError
    java.lang.IllegalStateException: Stream was terminated by error, no further calls are allowed
    ```
5. meanwhile, the onClose handler gets invoked because of the first error. As I suspected, there is a race condition here, we were able to process the second event before unregistering.
    ```
    INFO  [grpc-default-worker-ELG-3-10] 2022-01-20 17:22:44,663 Unregister after error
    ```

So the root issue is indeed at step 2, I'll fix that.
But because of the race condition I need to keep `isFailed`.

Reply:
Very good detective work @olim7t!

Will be interesting to know how the issue can be resolved: perhaps metadata about Keyspace, in context of DROP-Table, would only need Keyspace name? And not actual properties.
(or maybe nothing else would be included but processing will try to construct `Keyspace` object with metadata).

Reply:
Hmm from client perspective if something is wrong internally on the server, I am not sure if I would like to get that error back.. Yes you say it's client responsibility to reconnect, but from the client perspective I just want to listen on schema changes.. So actually I don't except any errors reported back, except for the "protocol" ones.. Can we re-think this and actually not call `onError` here at all.. Makes no sense from the client point of view..

Reply:
> Hmm from client perspective if something is wrong internally on the server, I am not sure if I would like to get that error back..

That was my initial approach in #1403. But the problem is that the client is now operating on a potentially stale version of the schema, without any indication that something went wrong. Propagating the error gives it a chance to start over from a fresh copy of the schema (see `SchemaCache.Observer.onError`).